pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,288 @@
+            TransformFunction dimJoinKeyFunction = joinArguments.get((i*2));
+            Preconditions.checkArgument(dimJoinKeyFunction instanceof LiteralTransformFunction,
+                "JoinKey argument must be a literal(string) representing the primary key for the dimension table");
+            _joinKeys.add(((LiteralTransformFunction) dimJoinKeyFunction).getLiteral());
+
+            TransformFunction factJoinValueFunction = joinArguments.get((i*2)+1);
+            TransformResultMetadata factJoinValueFunctionResultMetadata = factJoinValueFunction.getResultMetadata();
+            Preconditions.checkArgument(factJoinValueFunctionResultMetadata.isSingleValue(),
+                "JoinValue argument must be a single value expression");
+            _joinValueFunctions.add(factJoinValueFunction);
+        }
+
+        // Validate lookup table and relevant columns
+        _dataManager = DimensionTableDataManager.getInstanceByTableName(_dimTableName + TABLE_NAME_SUFFIX);
+        Preconditions.checkArgument(_dataManager != null,
+                String.format("Dimension table does not exist: %s", _dimTableName));
+
+        _lookupColumnFieldSpec = _dataManager.getColumnFieldSpec(_dimColumnName);
+        Preconditions.checkArgument(_lookupColumnFieldSpec != null,
+                String.format("Column does not exist in dimension table: %s:%s", _dimTableName, _dimColumnName));
+
+        for (String joinKey: _joinKeys) {
+            FieldSpec pkColumnSpec = _dataManager.getColumnFieldSpec(joinKey);
+            Preconditions.checkArgument(pkColumnSpec != null,
+                    String.format("Primary key column doesn't exist in dimension table: %s:%s",_dimTableName, joinKey));
+            _joinValueFieldSpecs.add(pkColumnSpec);
+        }
+    }
+
+    @Override
+    public TransformResultMetadata getResultMetadata() {
+        return new TransformResultMetadata(_lookupColumnFieldSpec.getDataType(),
+            _lookupColumnFieldSpec.isSingleValueField(),
+            false);
+    }
+
+    private Object[] lookup(ProjectionBlock projectionBlock) {
+        int numPkColumns = _joinKeys.size();
+        int numDocuments = projectionBlock.getNumDocs();
+        Object[][] pkColumns = new Object[numPkColumns][];
+        for (int i=0; i<numPkColumns; i++) {
+            FieldSpec.DataType colType = _joinValueFieldSpecs.get(i).getDataType();
+            TransformFunction tf = _joinValueFunctions.get(i);
+            switch (colType) {
+                case STRING:
+                    pkColumns[i] = tf.transformToStringValuesSV(projectionBlock);
+                    break;
+                case INT:
+                    pkColumns[i] = ArrayUtils.toObject(tf.transformToIntValuesSV(projectionBlock));
+                    break;
+                case LONG:
+                    pkColumns[i] = ArrayUtils.toObject(tf.transformToLongValuesSV(projectionBlock));
+                    break;
+                default:
```

Comment:
Do we plan to support other types like Double and Bytes?

Reply:
This is also my question. I couldn't find a definitive answer in existing code which values we support for PK, but STRING, INT, LONG made sense to me as a sensible subset. Let me know what you think @yupeng9 @Jackie-Jiang .

Reply:
You already have the implementation for other types in this class, so I think you can just add the support?